community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/IndexCursorProgressor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.newapi;
+
+import org.neo4j.internal.kernel.api.LabelSet;
+import org.neo4j.values.storable.Value;
+
+public interface IndexCursorProgressor extends AutoCloseable
```

Comment:
This interface together with nested counter parts could really use some javadoc. Without looking at the existing implementations it is very hard to figure out how they play together (and even after looking it sill gives me headaches 😅  ).

That being said, when you have figured out how those interfaces work they are pretty nice!